llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp

```diff
@@ -11898,6 +11864,12 @@ Value *BoUpSLP::vectorizeTree(
     Value *Vec = E->VectorizedValue;
     assert(Vec && "Can't find vectorizable value");
 
+    // Generate extract for in-tree uses if the use is scalar operand in
+    // vectorized instruction.
+    if (auto *UserVecTE = getTreeEntry(User))
+      if (doesInTreeUserNeedToExtract(Scalar, cast<Instruction>(User), TLI))
+        User = cast<llvm::User>(UserVecTE->VectorizedValue);
+
```

Comment:
Why need this? I meant try to fix the place where we buildingExtractUsers directly, I think it can be fixed by proper checking of the scalar element.

Reply:
> Why need this?

Adding new vectorized load/store/call that use in-tree scalar to `ExternalUses` in `BoUpSLP::vectorizeTree()` is removed, we still should generate extract for these new added vectorized load/store/call. As we add the original scalar load/store/call to `ExternalUses` in `buildingExtractUsers()`, so I add a check here, generate extract for new added vectorized load/store/call instead of the scalar load/store/call that will be erased later.

Take a llvm/test/Transforms/SLPVectorizer/X86/extract_in_tree_user.ll as an example.
```
define i32 @fn1() {
entry:
  %0 = load ptr, ptr @a, align 8
  %add.ptr = getelementptr inbounds i64, ptr %0, i64 11
  %1 = ptrtoint ptr %add.ptr to i64
  store i64 %1, ptr %add.ptr, align 8
  %add.ptr1 = getelementptr inbounds i64, ptr %0, i64 56
  %2 = ptrtoint ptr %add.ptr1 to i64
  %arrayidx2 = getelementptr inbounds i64, ptr %0, i64 12
  store i64 %2, ptr %arrayidx2, align 8
  ; store <2 x i64> %5, ptr %add.ptr, align 8
  ret i32 undef
}
```
Without this change,  when generating extracts, `ExternalUses` have
 `{%add.ptr, store i64 %1, ptr %add.ptr, align 8}` and
  `{%add.ptr, store <2 x i64> %5, ptr %add.ptr, align 8}`, 
what this patch want to do is only generate extract for `{%add.ptr, store <2 x i64> %5, ptr %add.ptr, align 8}` since there is no need to generate extract for `{%add.ptr, store i64 %1, ptr %add.ptr, align 8}`.

---

> I meant try to fix the place where we buildingExtractUsers directly, I think it can be fixed by proper checking of the scalar element.

Sorry, I don't quite understand this, could you please explain this a bit more?

Reply:
Ok, I see why you need this.
Instead, I suggest another approach. When building the external uses entry for such scalars, just mark that there is no exact user. In this case all uses of this scalar will be automatically replaced by extractelement and you don't need this change anymore.
But I think some adjustment in the original code for also should be done.
```
diff --git a/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp b/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp
index 5c325ad8a291..e66e2a7d110e 100644
--- a/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp
+++ b/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp
@@ -4928,33 +4928,33 @@ void BoUpSLP::buildExternalUses(
         if (!UserInst)
           continue;

-        if (isDeleted(UserInst))
-          continue;
-
         // Skip in-tree scalars that become vectors
         if (TreeEntry *UseEntry = getTreeEntry(U)) {
-          Value *UseScalar = UseEntry->Scalars[0];
           // Some in-tree scalars will remain as scalar in vectorized
-          // instructions. If that is the case, the one in Lane 0 will
+          // instructions. If that is the case, the one in FoundLane will
           // be used.
-          if (UseScalar != U ||
-              UseEntry->State == TreeEntry::ScatterVectorize ||
+          if (UseEntry->State == TreeEntry::ScatterVectorize ||
               UseEntry->State == TreeEntry::PossibleStridedVectorize ||
-              !doesInTreeUserNeedToExtract(Scalar, UserInst, TLI)) {
+              !doesInTreeUserNeedToExtract(
+                  Scalar, cast<Instruction>(UseEntry->Scalars.front()), TLI)) {
             LLVM_DEBUG(dbgs() << "SLP: \tInternal user will be removed:" << *U
                               << ".\n");
             assert(UseEntry->State != TreeEntry::NeedToGather && "Bad state");
             continue;
           }
+          U = nullptr;
         }

+        if (isDeleted(UserInst))
+          continue;
+
         // Ignore users in the user ignore list.
         if (UserIgnoreList && UserIgnoreList->contains(UserInst))
           continue;

         LLVM_DEBUG(dbgs() << "SLP: Need to extract:" << *U << " from lane "
                           << Lane << " from " << *Scalar << ".\n");
-        ExternalUses.push_back(ExternalUser(Scalar, U, FoundLane));
+        ExternalUses.emplace_back(Scalar, U, FoundLane);
       }
     }
   }

```